lib/libterm/termcap.3
.\"	$NetBSD: termcap.3,v 1.31 2005/05/15 21:11:13 christos Exp $
.\"
.\" Copyright (c) 1980, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)termcap.3	8.2 (Berkeley) 12/11/93
.\"
.Dd May 15, 2005
.Dt TERMCAP 3
.Os
.Sh NAME
.Nm tgetent ,
.Nm tgetnum ,
.Nm tgetflag ,
.Nm tgetstr ,
.Nm tgoto ,
.Nm tputs
.Nd terminal independent operation routines
.Sh LIBRARY
.Lb libtermcap
.Sh SYNOPSIS
.In termcap.h
.Vt char PC;
.Vt char *BC;
.Vt char *UP;
.Vt short ospeed;
.Vt struct tinfo *info;
.Ft int
.Fn tgetent "char *bp" "const char *name"
.Ft int
.Fn tgetnum "const char *id"
.Ft int
.Fn tgetflag "const char *id"
.Ft char *
.Fn tgetstr "const char *id" "char **area"
.Ft char *
.Fn tgoto "const char *cm" "int destcol" "int destline"
.Ft void
.Fn tputs "const char *cp" "int affcnt" "int (*outc)(int)"
.Ft int
.Fn t_getent "struct tinfo **info" "const char *name"
.Ft int
.Fn t_getnum "struct tinfo *info" "const char *id"
.Ft int
.Fn t_getflag "struct tinfo *info" "const char *id"
.Ft char *
.Fn t_getstr "struct tinfo *info" "const char *id" "char **area" "size_t *limit"
.Ft char *
.Fn t_agetstr "struct tinfo *info" "const char *id"
.Ft int
.Fn t_getterm "struct tinfo *info" "char **area" "size_t *limit"
.Ft int
.Fn t_goto "struct tinfo *info" "const char *id" "int destcol" "int destline" "char *buffer" "size_t limit"
.Ft int
.Fn t_puts "struct tinfo *info" "const char *cp" "int affcnt" "void (*outc)(char, void *)" "void *args"
.Ft void
.Fn t_freent "struct tinfo *info"
.Ft int
.Fn t_setinfo "struct tinfo **info" "const char *entry"
.Sh DESCRIPTION
These functions extract and use capabilities from a terminal capability data
base, usually
.Pa /usr/share/misc/termcap ,
the format of which is described in
.Xr termcap 5 .
These are low level routines;
see
.Xr curses 3
for a higher level package.
.Pp
The
.Fn tgetent
function
extracts the entry for terminal
.Fa name
into the buffer at
.Fa bp .
The
.Fa bp
argument
should be a character buffer of size
1024 and must be retained through all subsequent calls to
.Fn tgetnum ,
.Fn tgetflag ,
and
.Fn tgetstr .
The
.Fn tgetent
function
returns \-1 if none of the
.Nm termcap
data base files could be opened,
0 if the terminal name given does not have an entry,
and 1 if all goes well.
It will look in the environment for a
.Ev TERMCAP
variable.
If found, and the value does not begin with a slash, the value does
not contain the ZZ capability (see
.Em NOTES
for a description of this capability),
and the terminal type
.Fa name
is the same as the environment string
.Ev TERM ,
the
.Ev TERMCAP
string is used instead of reading a
.Nm termcap
file.
If the value does contain the ZZ capability then the
.Ev TERM
environment string is used to read
.Nm termcap ,
if the read fails for any reason the value of
.Ev TERMCAP
will be used despite it containing ZZ.
If
.Ev TERMCAP
does begin with a slash, the string is used as a path name
of the
.Nm termcap
file to search.
If
.Ev TERMCAP
does not begin with a slash and
.Fa name
is different from
.Ev TERM ,
.Fn tgetent
searches the files
.Pa $HOME/.termcap
and
.Pa /usr/share/misc/termcap ,
in that order, unless the environment variable
.Ev TERMPATH
exists,
in which case it specifies a list of file pathnames
(separated by spaces or colons) to be searched instead.
Whenever multiple files are searched and a
.Sy tc
field occurs in the requested entry, the entry it names must be found
in the same file or one of the succeeding files.
This can speed up entry into programs that call
.Fn tgetent ,
as well as help debug new terminal descriptions
or make one for your terminal if you can't write the file
.Pa /usr/share/misc/termcap .
.Pp
The
.Fn tgetnum
function
gets the numeric value of capability
.Fa id ,
returning \-1 if it is not given for the terminal.
The
.Fn tgetflag
function
returns 1 if the specified capability is present in
the terminal's entry, 0 if it is not.
The
.Fn tgetstr
function
returns the string value of the capability
.Fa id ,
places it in the buffer at
.Fa area ,
and advances the
.Fa area
pointer.
It decodes the abbreviations for this field described in
.Xr termcap 5 ,
except for cursor addressing and padding information.
The
.Fn tgetstr
function
returns
.Dv NULL
if the capability was not found.
.Pp
The
.Fn tgoto
function
returns a cursor addressing string decoded from
.Fa cm
to go to column
.Fa destcol
in line
.Fa destline .
It uses the external variables
.Va UP
(from the
.Sy up
capability)
and
.Va BC
(if
.Sy bc
is given rather than
.Sy bs )
if necessary to avoid placing
.Sy \en ,
.Sy ^D
or
.Sy ^@
in
the returned string.
(Programs which call
.Fn tgoto
should be sure to turn off the
.Dv XTABS
bit(s),
since
.Fn tgoto
may now output a tab.
Note that programs using termcap should in general turn off
.Dv XTABS
anyway since some terminals use control-I for other functions,
such as nondestructive space.)
If a
.Sy %
sequence is given which is not understood, then
.Fn tgoto
returns
.Pq Dv OOPS .
.Pp
The
.Fn tputs
function
decodes the leading padding information of the string
.Fa cp ;
.Fa affcnt
gives the number of lines affected by the operation, or 1 if this is
not applicable,
.Fa outc
is a routine which is called with each character in turn.
The external variable
.Va ospeed
should contain the output speed of the terminal as encoded by
.Xr stty 3 .
The external variable
.Va PC
should contain a pad character to be used (from the
.Sy pc
capability)
if a null
.Pq Sy ^@
is inappropriate.
.Pp
The
.Fn t_getent
function operates in a similar manner to the
.Fn tgetent
function excepting that the
.Fa info
argument is a pointer to a pointer of the opaque type
.Va tinfo .
If the call to
.Fn t_getent
succeeds then the argument
.Fa info
will be updated with the address of an object that contains the termcap
entry.
This pointer can then be passed to calls of
.Fn t_getnum ,
.Fn t_getflag
and
.Fn t_getstr .
When the information pointed to by
.Fa info
is no longer required any storage associated with the object can be
released by calling
.Fn t_freent .
.Pp
The functions
.Fn t_getnum
and
.Fn t_getflag
operate in the same manner as
.Fn tgetnum
and
.Fn tgetflag
with the exception that the pointer to the termcap object is passed along
with the id of the capability required.
.Pp
The function
.Fn t_getstr
performs the same function as
.Fn tgetstr
but has a
.Fa limit
parameter that gives the number of characters that can be inserted in to
the array pointed to by
.Fa area .
The
.Fa limit
argument is updated by the
.Fn t_getstr
call to give the number of characters that remain available in
.Fa area .
If the t_getstr call fails then
.Dv NULL
will be returned and errno set to indicate the failure,
.Er ENOENT
indicates there was no termcap entry for the given
.Fa id ,
.Er E2BIG
indicates the retrieved entry would have overflowed
.Fa area .
If t_getstr is called with
.Fa area
being
.Dv NULL
then the size required to hold the capability string will be returned in
.Fa limit
so the caller can allocate enough storage to hold the capability.
.Pp
The function
.Fn t_agetstr
performs the same function as
.Fn t_getstr
except it handles memory allocation automatically.
The memory that
.Fn t_agetstr
allocates will be freed when
.Fn t_freent
is called.
.Pp
The function
.Fn t_getterm
returns a copy of the termcap name string of the termcap entry
associated with
.Fa info
in the buffer pointed to by
.Fa area .
.Fn t_getterm
returns 0 on success and \-1 on error.
On error errno will be set to
.Er EINVAL
if the termcap entry in
.Fa info
is malformed or
.Er E2BIG
if the size of the name exceeds the size specified by
.Fa limit .
If
.Fa area
is
.Dv NULL
then the size required to hold the terminal name will be returned in
.Fa limit
allowing sufficient storage to be allocated.
If
.Fa limit
is
.Dv NULL
then no bounds checking will be performed.
.Pp
The
.Fn t_goto
function is the same as the
.Fn tgoto
function excepting that the capabilities for
.Sy up
and
.Sy bc
are extracted from the
.Fa info
object and that the string formed by
.Fn t_goto
is placed in the
.Fa buffer
argument, the number of characters allowed to be placed in
.Fa buffer
is controlled by
.Fa limit .
If the expansion performed by
.Fn t_goto
would exceed the space in
.Fa buffer
then
.Fn t_goto
will return \-1 and set errno to
.Er E2BIG .
The function
.Fn t_puts
is similar to the
.Fn tputs
function excepting that
.Fa info
holds a pointer to the termcap object that was returned by a previous
.Fn t_getent
call, this object will be used to retrieve the
.Sy pc
attribute for the terminal.
The
.Fa outc
function is a pointer to a function that will be called by
.Fn t_puts
to output each character in the
.Fa cp
string.
The
.Fa outc
function will be called with two parameters.
The first is the character
to be printed and the second is an optional argument that was passed to
.Fn t_puts
in the
.Fa args
argument.
The interpretation of the contents of
.Fa args
is dependent solely on the implementation of
.Fa outc .
.Pp
The
.Fn t_setinfo
function allows the termcap entry contained in the
.Fa entry
string to be inserted into the
.Fa info
structure.
Memory sufficient to hold the contents of
.Fa entry
is automatically allocated.
This allows the programmer to provide a fail over terminal capability
string if fetching the termcap entry from the termcap database fails.
The format of the string
.Fa entry
is assumed to be a valid termcap entry.
.Pp
NOTE: A special capability of
.Fa ZZ
is added to the end of the termcap entry retrieved.
The number that follows this entry is the address of the buffer allocated
to hold the full termcap entry.
The caller may retrieve the pointer to the extended buffer by performing a
.Fn tgetstr
to retrieve the
.Fa ZZ
capability, the string is the output of a
.Fn printf
%p and may be converted back to a pointer using
.Fn sscanf
or similar.
The ZZ capability is only necessary if the caller wishes to
directly manipulate the termcap entry, all the termcap function calls
automatically use the extended buffer to retrieve terminal capabilities.
.Sh FILES
.Bl -tag -width /usr/share/misc/termcap -compact
.It Pa /usr/lib/libtermcap.a
.Fl l Ar termcap
library (also known as
.Fl l Ar termlib )
.It Pa /usr/share/misc/termcap
standard terminal capability data base
.It Pa $HOME/.termcap
user's terminal capability data base
.El
.Sh SEE ALSO
.Xr ex 1 ,
.Xr curses 3 ,
.Xr termcap 5
.Sh HISTORY
The
.Nm termcap
t_*() functions appeared in
.Nx 1.5 .
The rest of the
.Nm termcap
functions appeared in
.Bx 4.0 .


lib/libterm/termcap.c
/*	$NetBSD: termcap.c,v 1.49 2006/03/18 12:18:15 blymn Exp $	*/

/*
 * Copyright (c) 1980, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#ifndef lint
#if 0
static char sccsid[] = "@(#)termcap.c	8.1 (Berkeley) 6/4/93";
#else
__RCSID("$NetBSD: termcap.c,v 1.49 2006/03/18 12:18:15 blymn Exp $");
#endif
#endif /* not lint */

#include <sys/types.h>
#include <sys/param.h>
#include <assert.h>
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <termcap.h>
#include <errno.h>
#include "pathnames.h"
#include "termcap_private.h"

#define	PBUFSIZ		MAXPATHLEN	/* max length of filename path */
#define	PVECSIZ		32		/* max number of names in path */

/*
 * termcap - routines for dealing with the terminal capability data base
 *
 * BUG:		Should use a "last" pointer in tbuf, so that searching
 *		for capabilities alphabetically would not be a n**2/2
 *		process when large numbers of capabilities are given.
 * Note:	If we add a last pointer now we will screw up the
 *		tc capability. We really should compile termcap.
 *
 * Essentially all the work here is scanning and decoding escapes
 * in string capabilities.  We don't use stdio because the editor
 * doesn't, and because living w/o it is not hard.
 */

static char *tbuf = NULL;		/* termcap buffer */
static struct tinfo *fbuf = NULL;	/* untruncated termcap buffer */

/*
 * Set the termcap entry to the arbitrary string passed in, this can
 * be used to provide a "dummy" termcap entry if a real one does not
 * exist.  This function will malloc the buffer and space for the
 * string.  If an error occurs return -1 otherwise return 0.
 */
int
t_setinfo(struct tinfo **bp, const char *entry)
{
	char capability[256], *cap_ptr;
	size_t limit;

	_DIAGASSERT(bp != NULL);
	_DIAGASSERT(entry != NULL);

	if ((*bp = malloc(sizeof(struct tinfo))) == NULL)
		return -1;

	if (((*bp)->info = (char *) malloc(strlen(entry) + 1)) == NULL)
		return -1;

	strcpy((*bp)->info, entry);

	cap_ptr = capability;
	limit = 255;
	(*bp)->up = t_getstr(*bp, "up", &cap_ptr, &limit);
	if ((*bp)->up)
		(*bp)->up = strdup((*bp)->up);
	cap_ptr = capability;
	limit = 255;
	(*bp)->bc = t_getstr(*bp, "bc", &cap_ptr, &limit);
	if ((*bp)->bc)
		(*bp)->bc = strdup((*bp)->bc);
	(*bp)->tbuf = NULL;

	return 0;
}

/*
 * Get an extended entry for the terminal name.  This differs from
 * tgetent only in a) the buffer is malloc'ed for the caller and
 * b) the termcap entry is not truncated to 1023 characters.
 */

int
t_getent(struct tinfo **bp, const char *name)
{
	char  *p;
	char  *cp;
	char **fname;
	char  *home;
	int    i, did_getset;
	size_t limit;
	char   pathbuf[PBUFSIZ];	/* holds raw path of filenames */
	char  *pathvec[PVECSIZ];	/* to point to names in pathbuf */
	char  *termpath;
	char  capability[256], *cap_ptr;
	int error;


	_DIAGASSERT(bp != NULL);
	_DIAGASSERT(name != NULL);

	if ((*bp = malloc(sizeof(struct tinfo))) == NULL)
		return 0;

	fname = pathvec;
	p = pathbuf;
	cp = getenv("TERMCAP");
	/*
	 * TERMCAP can have one of two things in it. It can be the
	 * name of a file to use instead of
	 * /usr/share/misc/termcap. In this case it better start with
	 * a "/". Or it can be an entry to use so we don't have to
	 * read the file. In this case cgetset() withh crunch out the
	 * newlines.  If TERMCAP does not hold a file name then a path
	 * of names is searched instead.  The path is found in the
	 * TERMPATH variable, or becomes _PATH_DEF ("$HOME/.termcap
	 * /usr/share/misc/termcap") if no TERMPATH exists.
	 */
	if (!cp || *cp != '/') {	/* no TERMCAP or it holds an entry */
		if ((termpath = getenv("TERMPATH")) != NULL)
			(void)strlcpy(pathbuf, termpath, sizeof(pathbuf));
		else {
			if ((home = getenv("HOME")) != NULL) {
				/* set up default */
				p += strlen(home);	/* path, looking in */
				(void)strlcpy(pathbuf, home,
				    sizeof(pathbuf)); /* $HOME first */
				if ((size_t)(p - pathbuf) < sizeof(pathbuf) - 1)
				    *p++ = '/';
			}	/* if no $HOME look in current directory */
			if ((size_t)(p - pathbuf) < sizeof(pathbuf) - 1) {
			    (void)strlcpy(p, _PATH_DEF,
				sizeof(pathbuf) - (p - pathbuf));
			}
		}
	}
	else {
		/* user-defined name in TERMCAP; still can be tokenized */
		(void)strlcpy(pathbuf, cp, sizeof(pathbuf));
	}

	*fname++ = pathbuf;	/* tokenize path into vector of names */
	while (*++p)
		if (*p == ' ' || *p == ':') {
			*p = '\0';
			while (*++p)
				if (*p != ' ' && *p != ':')
					break;
			if (*p == '\0')
				break;
			*fname++ = p;
			if (fname >= pathvec + PVECSIZ) {
				fname--;
				break;
			}
		}
	*fname = NULL;			/* mark end of vector */

	/*
	 * try ignoring TERMCAP if it has a ZZ in it, we do this
	 * because a TERMCAP with ZZ in it indicates the entry has been
	 * exported by another program using the "old" interface, the
	 * termcap entry has been truncated and ZZ points to an address
	 * in the exporting programs memory space which is of no use
	 * here - anyone who is exporting the termcap entry and then
	 * reading it back again in the same program deserves to be
	 * taken out, beaten up, dragged about, shot and then hurt some
	 * more.
	 */
	did_getset = 0;
	if (cp && *cp && *cp != '/' && strstr(cp, ":ZZ") == NULL) {
		did_getset = 1;
		if (cgetset(cp) < 0) {
			error = -2;
			goto out;
		}
	}

	/*
	 * XXX potential security hole here in a set-id program if the
	 * user had setup name to be built from a path they can not
	 * normally read.
	 */
 	(*bp)->info = NULL;
 	i = cgetent(&((*bp)->info), pathvec, name);

	/*
	 * if we get an error and we skipped doing the cgetset before
	 * we try with TERMCAP in place - we may be using a truncated
	 * termcap entry but what else can one do?
	 */
	if ((i < 0) && (did_getset == 0)) {
		if (cp && *cp && *cp != '/')
			if (cgetset(cp) < 0) {
				error = -2;
				goto out;
			}
		i = cgetent(&((*bp)->info), pathvec, name);
	}

	/* no tc reference loop return code in libterm XXX */
	if (i == -3) {
		error = -1;
		goto out;
	}

	/*
	 * fill in t_goto capabilities - this prevents memory leaks
	 * and is more efficient than fetching these capabilities
	 * every time t_goto is called.
	 */
	if (i >= 0) {
		cap_ptr = capability;
		limit = 255;
		(*bp)->up = t_getstr(*bp, "up", &cap_ptr, &limit);
		if ((*bp)->up)
			(*bp)->up = strdup((*bp)->up);
		cap_ptr = capability;
		limit = 255;
		(*bp)->bc = t_getstr(*bp, "bc", &cap_ptr, &limit);
		if ((*bp)->bc)
			(*bp)->bc = strdup((*bp)->bc);
		(*bp)->tbuf = NULL;
	} else {
		error = i + 1;
		goto out;
	}

	return (i + 1);
out:
	free(*bp);
	*bp = NULL;
	return error;
}

/*
 * Get an entry for terminal name in buffer bp from the termcap file.
 */
int
tgetent(char *bp, const char *name)
{
	int i, plen, elen, c;
        char *ptrbuf = NULL;

	i = t_getent(&fbuf, name);

	if (i == 1) {
		/*
		 * stash the full buffer pointer as the ZZ capability
		 * in the termcap buffer passed.
		 */
                plen = asprintf(&ptrbuf, ":ZZ=%p", fbuf->info);
		(void)strlcpy(bp, fbuf->info, 1024);
                elen = strlen(bp);
		/*
		 * backup over the entry if the addition of the full
		 * buffer pointer will overflow the buffer passed.  We
		 * want to truncate the termcap entry on a capability
		 * boundary.
		 */
                if ((elen + plen) > 1023) {
			bp[1023 - plen] = '\0';
			for (c = (elen - plen); c > 0; c--) {
				if (bp[c] == ':') {
					bp[c] = '\0';
					break;
				}
			}
		}

		strcat(bp, ptrbuf);
                tbuf = bp;
	}

	return i;
}

/*
 * Return the (numeric) option id.
 * Numeric options look like
 *	li#80
 * i.e. the option string is separated from the numeric value by
 * a # character.  If the option is not found we return -1.
 * Note that we handle octal numbers beginning with 0.
 */
int
t_getnum(struct tinfo *info, const char *id)
{
	long num;

	_DIAGASSERT(info != NULL);
	_DIAGASSERT(id != NULL);

	if (cgetnum(info->info, id, &num) == 0)
		return (int)(num);
	else
		return (-1);
}

int
tgetnum(const char *id)
{
	return fbuf ? t_getnum(fbuf, id) : -1;
}

/*
 * Handle a flag option.
 * Flag options are given "naked", i.e. followed by a : or the end
 * of the buffer.  Return 1 if we find the option, or 0 if it is
 * not given.
 */
int t_getflag(struct tinfo *info, const char *id)
{
	_DIAGASSERT(info != NULL);
	_DIAGASSERT(id != NULL);

	return (cgetcap(info->info, id, ':') != NULL);
}

int
tgetflag(const char *id)
{
	return fbuf ? t_getflag(fbuf, id) : 0;
}

/*
 * Get a string valued option.
 * These are given as
 *	cl=^Z
 * Much decoding is done on the strings, and the strings are
 * placed in area, which is a ref parameter which is updated.
 * limit is the number of characters allowed to be put into
 * area, this is updated.
 */
char *
t_getstr(struct tinfo *info, const char *id, char **area, size_t *limit)
{
	char *s;
	int i;

	_DIAGASSERT(info != NULL);
	_DIAGASSERT(id != NULL);
	/* area may be NULL */


	if ((i = cgetstr(info->info, id, &s)) < 0) {
		errno = ENOENT;
		if ((area == NULL) && (limit != NULL))
			*limit = 0;
		return NULL;
	}

	i ++; /* account for the NUL */
	if (area != NULL) {
		/*
		 * check if there is room for the new entry to be put into
		 * area
		 */
		if (limit != NULL && (*limit < (size_t) i)) {
			errno = E2BIG;
			free(s);
			return NULL;
		}

		(void)strcpy(*area, s);
		free(s);
		s = *area;
		*area += i;
		if (limit != NULL) *limit -= i;

		return (s);
	} else {
		_DIAGASSERT(limit != NULL);
		*limit = i;
		free(s);
		return NULL;
	}
}

/*
 * Get a string valued option.
 * These are given as
 *	cl=^Z
 * Much decoding is done on the strings, and the strings are
 * placed in area, which is a ref parameter which is updated.
 * No checking on area overflow.
 */
char *
tgetstr(const char *id, char **area)
{
	struct tinfo dummy;
	char ids[3];

	_DIAGASSERT(id != NULL);

	if (fbuf == NULL)
		return NULL;

	/*
	 * XXX
	 * This is for all the boneheaded programs that relied on tgetstr
	 * to look only at the first 2 characters of the string passed...
	 */
	ids[0] = id[0];
	ids[1] = id[1];
	ids[2] = '\0';

	if ((id[0] == 'Z') && (id[1] == 'Z')) {
		dummy.info = tbuf;
		return t_getstr(&dummy, ids, area, NULL);
	}
	else
		return t_getstr(fbuf, ids, area, NULL);
}

/*
 * Return a string valued option specified by id, allocating memory to
 * an internal buffer as necessary. The memory allocated can be
 * free'd by a call to t_freent().
 *
 * If the string is not found or memory allocation fails then NULL
 * is returned.
 */
#define BSIZE 256
char *
t_agetstr(struct tinfo *info, const char *id)
{
	size_t new_size;
	struct tbuf *tb;

	_DIAGASSERT(info != NULL);
	_DIAGASSERT(id != NULL);

	t_getstr(info, id, NULL, &new_size);

	/* either the string is empty or the capability does not exist. */
	if (new_size == 0)
		return NULL;

	if ((tb = info->tbuf) == NULL ||
	    (size_t) (tb->eptr - tb->ptr) < (new_size + 1)) {
		if (new_size < BSIZE)
			new_size = BSIZE;
		else
			new_size++;

		if ((tb = malloc(sizeof(*info->tbuf))) == NULL)
			return NULL;

		if ((tb->data = tb->ptr = tb->eptr = malloc(new_size))
		    == NULL) {
			free(tb);
			return NULL;
		}

		tb->eptr += new_size;

		if (info->tbuf != NULL)
			tb->next = info->tbuf;
		else
			tb->next = NULL;

		info->tbuf = tb;
	}
	return t_getstr(info, id, &tb->ptr, NULL);
}

/*
 * Free the buffer allocated by t_getent
 *
 */
void
t_freent(struct tinfo *info)
{
	struct tbuf *tb, *wb;
	_DIAGASSERT(info != NULL);
	free(info->info);
	if (info->up != NULL)
		free(info->up);
	if (info->bc != NULL)
		free(info->bc);
	for (tb = info->tbuf; tb;) {
		wb = tb;
		tb = tb->next;
		free(wb->data);
		free(wb);
	}
	free(info);
}

/*
 * Get the terminal name string from the termcap entry.
 *
 */
int
t_getterm(struct tinfo *info, char **area, size_t *limit)
{
	char *endp;
	size_t count;

	_DIAGASSERT(info != NULL);
	if ((endp = strchr(info->info, ':')) == NULL) {
		errno = EINVAL;
		return -1;
	}


	count = endp - info->info + 1;
	if (area == NULL) {
		_DIAGASSERT(limit != NULL);
		*limit = count;
		return 0;
	} else {
		if ((limit != NULL) && (count > *limit)) {
			errno = E2BIG;
			return -1;
		}

		(void)strlcpy(*area, info->info, count);
		if (limit != NULL)
			*limit -= count;
	}

	return 0;
}
